### DISCUSSION TITLE
update/expire identity map with upsert

### DISCUSSION THREAD

**Discussion by ay-ramp** [sqlalchemy/sqlalchemy#9306] [Usage Questions] [answered]
hi there!

we're using a few postgres upserts in various places like this:

```
def upsert_entity(values: ...) -> int:
   """Upsert the Entity and return its ID"""
    insert_stmt = (
        insert(Entity)
        .values(values)
        .returning(Entity.id)
    )
    do_update_stmt = insert_stmt.on_conflict_do_update(
        index_elements=["foo"],
        set_={key: getattr(insert_stmt.excluded, key) for key in values.keys()},
    )

    entity_id = session.scalar(
        do_update_stmt
    )
    return entity_id
```

this works great, with the exception that it doesn't update the identity map for the object, so if `Entity` is fetched (by another function) before the upsert, and then subsequently fetched (by yet another function) after the upsert, the correct values don't get updated.

i saw that `synchronize_session` might be helpful, but per the [note](https://docs.sqlalchemy.org/en/14/orm/session_basics.html#orm-expression-update-delete) it doesn't seem to work well in combination with `returning`, which we need to get the ID. 

right now, i'm doing what feels like a pretty hacky manual expiration of the identity map:

```
def upsert_entity(values: ...) -> int:
   """Upsert the Entity and return its ID"""
    insert_stmt = (
        insert(Entity)
        .values(values)
        .returning(Entity.id)
    )
    do_update_stmt = insert_stmt.on_conflict_do_update(
        index_elements=["foo"],
        set_={key: getattr(insert_stmt.excluded, key) for key in values.keys()},
    )

    entity_id = session.scalar(
        do_update_stmt
    )

    key = identity_key(Entity, entity_id)
    if (instance := db.session.identity_map.get(key)) is not None:
        db.session.expire(instance)

    return entity_id
```

which seems to work but feels fragile.

would you mind sharing if that's the correct approach, or if we should be doing something else?

**Comment by zzzeek** [comment, marked as answer]
if you're on 2.0 and all that, I think [populate_existing](https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#orm-queryguide-populate-existing) (and also RETURN the full `Entity`) would make sure those RETURNING objects overwrite existing ones.

all very new use cases so let me know if that solves the issue (it should)

**Reply by zzzeek** [reply]
oh, you would want to return the full entity for that to work :)  sorry forgot that part

**Reply by zzzeek** [reply]
there is also a notion of a feature we might add for this someday which is #8070 .

**Reply by ay-ramp** [reply]
i'm stuck on 1.4 for the foreseeable future unfortunately

**Reply by zzzeek** [reply]
oh, then do what you're doing

**Reply by ay-ramp** [reply]
sounds good. thanks for the confirmation!

looking forward to first-class upsert landing one day :)

**Reply by CaselIT** [reply]
Can't insert be placed inside select using `from_statement` in 1.4? IIRC session understands this

```py
insert = insert(...).returning(Entity) # not only the id
query = select(Entity).from_statement(insert).execution_options(populate_existing=True)

session.execute(query)
```

**Reply by zzzeek** [reply]
ah yes, you can do that.  I guess that works here
